File: python/flintdb/crypto.py
class Crypto:
    """Utility class for handling all cryptographic operations."""
    @staticmethod
    def kek(kek: str) -> bytes | None:
        """Normalizes kek for use with Fernet.

        Returns:
            None if kek is empty, base64 encoded sha256 of kek otherwise.
        """
        if not kek:
            return None

        import hashlib, base64
        sha256_object = hashlib.sha256(kek.encode("utf-8"))
        return base64.b64encode(sha256_object.digest())

    @staticmethod
    def random_dek(kek: bytes) -> str:
        """Randomly generates Data Encryption Key (DEK).

        Args:
            kek: The Key Encryption Key (KEK) to encrypt the DEK with.

        Returns:
            Generated and encrypted Data Encryption Key (DEK).
        """
        from cryptography.fernet import Fernet
        dek = Fernet.generate_key().decode("utf-8")
        return Crypto.encrypt(dek, kek)

    @staticmethod
    def encrypt(data: "Any", kek: bytes) -> str:
        """Encrypts data using provided KEK.

        Args:
            data: The data to encrypt.
            kek: The key to be used for encryption.

        Returns:
            The ciphertext.
        """
        import json
        from cryptography.fernet import Fernet

        fernet = Fernet(kek)
        json_data = json.dumps(data).encode("utf-8")
        value = fernet.encrypt(json_data)
        return value.decode("utf-8")

    @staticmethod
    def decrypt(cipher: str, kek: bytes) -> "Any":
        """Decrypts a cipher string back to into data.

        Args:
            cipher: The encrypted data string.
            kek: The key used for decryption.

        Returns:
            The original data.
        """
        import json
        from cryptography.fernet import Fernet, InvalidToken

        fernet = Fernet(kek)
        value = fernet.decrypt(data.encode("utf-8"))
        return json.loads(value.decode("utf-8"))

    @staticmethod
    def random_id(nbytes: int | None = None) -> str:
        """Returns random string of nbytes*2 size."""
        import secrets
        return secrets.token_hex(nbytes)

    @staticmethod
    def hash(data: "Any") -> str:
        """Returns string hash of data."""
        import pickle, xxhash
        bytes = pickle.dumps(data)
        hash_object = xxhash.xxh64(bytes)
        return hash_object.hexdigest()

File: php/flintdb/Crypto.php
<?php

namespace FlintDB;

class Crypto {
  /**
   * Cipher method
   * @var string
   */
  const METHOD = 'AES-256-CBC';
  
  /**
   * Returns encrypted data
   * @param mixed $data
   * @param string $phrase
   * @return string
   */
  public static function encrypt( mixed $data, #[\SensitiveParameter] string $phrase ): string {
    $data = json_encode( $data );
    $iv_len = openssl_cipher_iv_length( self::METHOD );
    $iv = openssl_random_pseudo_bytes( $iv_len );
    $cipher = openssl_encrypt( $data, self::METHOD, $phrase, OPENSSL_RAW_DATA, $iv );
    $hmac = hash_hmac( 'sha256', $cipher, $phrase, true );
    return base64_encode( $iv . $hmac . $cipher );
  }
  
  /**
   * Returns decrypted data
   * @param string $hash
   * @param string $phrase
   * @return mixed
   */
  public static function decrypt( string $hash, #[\SensitiveParameter] string $phrase ): mixed {
    $hash = base64_decode( $hash );
    $iv_len = openssl_cipher_iv_length( self::METHOD );
    $iv = substr( $hash, 0, $iv_len );
    $hmac = substr( $hash, $iv_len, 32 );
    $ciphertext = substr( $hash, $iv_len + 32 );
    $original = openssl_decrypt( $ciphertext, self::METHOD, $phrase, OPENSSL_RAW_DATA, $iv );
    $calcmac = hash_hmac( 'sha256', $ciphertext, $phrase, true );
    if ( hash_equals( $hmac, $calcmac ) ) {
      return json_decode( $original, true );
    }
    
    return null;
  }
  
  /**
   * Returns random hex
   * @param int $nbytes
   * @return string
   */
  public static function random_id( int $nbytes = 32 ): string {
    return bin2hex( random_bytes( $nbytes ) );
  }
}
